persistence/rocksdb/src/main/java/org/infinispan/persistence/rocksdb/RocksDBStore.java

```diff
@@ -634,4 +516,580 @@ public RocksEntryIterator(RocksIterator it, Predicate<? super K> filter, boolean
+                // Force a GC to ensure that open file handles are released in Windows
+                System.gc();
+                String dataLocation = getQualifiedLocation();
+                Util.recursiveFileRemove(new File(dataLocation));
+                db = open(getQualifiedLocation(), dataDbOptions());
+
+                String expirationLocation = getQualifiedExpiredLocation();
+                Util.recursiveFileRemove(new File(expirationLocation));
+                expiredDb = openDatabase(expirationLocation, expiredDbOptions());
+            } finally {
+                semaphore.release(Integer.MAX_VALUE);
+            }
+        }
+
+        protected RocksIterator wrapIterator(RocksDB db, int segment) {
+            // Some Cache Store tests use clear and in case of the Rocks DB implementation
+            // this clears out internal references and results in throwing exceptions
+            // when getting an iterator. Unfortunately there is no nice way to check that...
+            return db.newIterator(defaultColumnFamilyHandle, new ReadOptions().setFillCache(false));
+        }
+
+        @Override
+        Publisher<K> publishKeys(IntSet segments, Predicate<? super K> filter) {
+            Predicate<? super K> combinedFilter = PersistenceUtil.combinePredicate(segments, keyPartitioner, filter);
+            return publish(-1, it -> Flowable.fromIterable(() -> new RocksKeyIterator(it, combinedFilter)));
+        }
+
+        @Override
+        Publisher<MarshalledEntry<K, V>> publishEntries(IntSet segments, Predicate<? super K> filter, boolean fetchValue,
+              boolean fetchMetadata) {
+            Predicate<? super K> combinedFilter = PersistenceUtil.combinePredicate(segments, keyPartitioner, filter);
+            return publish(-1, it -> Flowable.fromIterable(() -> {
+                // Make sure this is taken when the iterator is created
+                long now = timeService.wallClockTime();
+                return new RocksEntryIterator(it, combinedFilter, fetchValue, fetchMetadata, now);
+            }));
+        }
+
+        @Override
+        void addSegments(IntSet segments) {
+            // Do nothing
+        }
+
+        @Override
+        void removeSegments(IntSet segments) {
+            clear(segments);
+        }
+    }
+
+    private class SegmentedRocksDBHandler extends RocksDBHandler {
+        private final KeyPartitioner keyPartitioner;
+        private final AtomicReferenceArray<ColumnFamilyHandle> handles;
+
+        private SegmentedRocksDBHandler(int segmentCount, KeyPartitioner keyPartitioner) {
+            this.keyPartitioner = keyPartitioner;
+            this.handles = new AtomicReferenceArray<>(segmentCount);
+        }
+
+        byte[] byteArrayFromInt(int val) {
+            byte[] b = new byte[4];
+            b[3] = (byte) (val);
+            b[2] = (byte) (val >>> 8);
+            b[1] = (byte) (val >>> 16);
+            b[0] = (byte) (val >>> 24);
+            return b;
```

Comment:
It doesn't really matter, but could be:

```java
return new byte[] {
            (byte) (val >>> 24),
            (byte) (val >>> 16),
            (byte) (val >>> 8),
            (byte) val
};
```

Reply:
Sure, this is just how the `Bits` class normally does it :) But I think this is a little easier to read.